### PR TITLE
Make the header list size limit configurable on both client and serve side.

### LINKED ISSUE #1144: Make the total headers size limit configurable.
We may want to have our own limit, but before we decide the number, should just remove the current limit.

### PR DESCRIPTION
**Pull request by madongfly**
Fixes #1144

@louiscryan @nmittler Please take a look, thanks!

### CODE REVIEW

**Comment by nmittler** [comment]
Is there no way of limiting this in OkHttp?

**Comment by madongfly** [comment]
No, OkHttp doesn't have such limit.

File: netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java [hunk 1]
```diff
@@ -69,6 +70,7 @@
   private SslContext sslContext;
   private int flowControlWindow = DEFAULT_FLOW_CONTROL_WINDOW;
   private int maxMessageSize = DEFAULT_MAX_MESSAGE_SIZE;
+  private int maxHeaderListSize = Http2CodecUtil.DEFAULT_MAX_HEADER_SIZE;
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java]: Let's make our own limit in GrpcUtil

File: netty/src/main/java/io/grpc/netty/NettyServerBuilder.java [hunk 2]
```diff
@@ -68,6 +69,7 @@
   private int maxConcurrentCallsPerConnection = Integer.MAX_VALUE;
   private int flowControlWindow = DEFAULT_FLOW_CONTROL_WINDOW;
   private int maxMessageSize = DEFAULT_MAX_MESSAGE_SIZE;
+  private int maxHeaderListSize = Http2CodecUtil.DEFAULT_MAX_HEADER_SIZE;
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/netty/NettyServerBuilder.java]: Same as previous comment.

**Comment by nmittler** [comment]
@madongfly thanks for taking care of this ... just one general comment that we should use our own limit rather than using Netty's constant.

File: netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java (hunk 1 above)
Comment (madongfly) [inline netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java]: Done.

File: netty/src/main/java/io/grpc/netty/NettyServerBuilder.java (hunk 2 above)
Comment (madongfly) [inline netty/src/main/java/io/grpc/netty/NettyServerBuilder.java]: Done.

**Comment by madongfly** [comment]
@nmittler PTAL, thanks!

File: netty/src/main/java/io/grpc/netty/NettyClientTransport.java [hunk 3]
```diff
@@ -244,7 +250,9 @@ private void notifyTerminated(Status status) {
 
   private NettyClientHandler newHandler() {
     Http2Connection connection = new DefaultHttp2Connection(false);
-    Http2FrameReader frameReader = new DefaultHttp2FrameReader();
+    Http2HeadersDecoder headersDecoder =
+        new DefaultHttp2HeadersDecoder(maxHeaderListSize, Http2CodecUtil.DEFAULT_HEADER_TABLE_SIZE);
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/netty/NettyClientTransport.java]: Use the new constant in gRPC.

File: netty/src/main/java/io/grpc/netty/NettyServerTransport.java [hunk 4]
```diff
@@ -133,8 +138,10 @@ private void notifyTerminated(Throwable t) {
   private NettyServerHandler createHandler(ServerTransportListener transportListener) {
     Http2Connection connection = new DefaultHttp2Connection(true);
     Http2FrameLogger frameLogger = new Http2FrameLogger(LogLevel.DEBUG, getClass());
-    Http2FrameReader frameReader =
-        new Http2InboundFrameLogger(new DefaultHttp2FrameReader(), frameLogger);
+    Http2HeadersDecoder headersDecoder =
+        new DefaultHttp2HeadersDecoder(maxHeaderListSize, Http2CodecUtil.DEFAULT_HEADER_TABLE_SIZE);
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/netty/NettyServerTransport.java]: Use the new constant in gRPC.

File: netty/src/test/java/io/grpc/netty/NettyClientTransportTest.java [hunk 5]
```diff
@@ -154,14 +156,15 @@ public void headersShouldOverrideDefaultUserAgent() throws Exception {
...
+        receivedHeaders.get(USER_AGENT_KEY));
   }
 
   @Test
   public void maxMessageSizeShouldBeEnforced() throws Throwable {
     startServer();
     // Allow the response payloads of up to 1 byte.
-    NettyClientTransport transport = newTransport(newNegotiator(), 1);
+    NettyClientTransport transport = newTransport(newNegotiator(),
+        1, Http2CodecUtil.DEFAULT_MAX_HEADER_SIZE);
```
Comment (nmittler) [inline netty/src/test/java/io/grpc/netty/NettyClientTransportTest.java]: same comment.

File: netty/src/test/java/io/grpc/netty/NettyClientTransportTest.java [hunk 6]
```diff
@@ -205,7 +208,7 @@ public void creatingMultipleTlsTransportsShouldSucceed() throws Exception {
   @Test
   public void bufferedStreamsShouldBeClosedWhenConnectionTerminates() throws Exception {
     // Only allow a single stream active at a time.
-    startServer(1);
+    startServer(1, Http2CodecUtil.DEFAULT_MAX_HEADER_SIZE);
```
Comment (nmittler) [inline netty/src/test/java/io/grpc/netty/NettyClientTransportTest.java]: and here.

File: netty/src/test/java/io/grpc/netty/NettyClientTransportTest.java [hunk 7]
```diff
@@ -242,28 +290,30 @@ private ProtocolNegotiator newNegotiator() throws IOException {
   }
 
   private NettyClientTransport newTransport(ProtocolNegotiator negotiator) {
-    return newTransport(negotiator, DEFAULT_MAX_MESSAGE_SIZE);
+    return newTransport(negotiator,
+        DEFAULT_MAX_MESSAGE_SIZE, Http2CodecUtil.DEFAULT_MAX_HEADER_SIZE);
```
Comment (nmittler) [inline netty/src/test/java/io/grpc/netty/NettyClientTransportTest.java]: and here.

File: netty/src/test/java/io/grpc/netty/NettyClientTransportTest.java [hunk 8]
```diff
@@ -242,28 +290,30 @@ private ProtocolNegotiator newNegotiator() throws IOException {
...
     NettyClientTransport transport = new NettyClientTransport(address, NioSocketChannel.class,
-            group, negotiator, DEFAULT_WINDOW_SIZE, maxMsgSize, authority);
+            group, negotiator, DEFAULT_WINDOW_SIZE, maxMsgSize, maxHeaderListSize, authority);
     transports.add(transport);
     return transport;
   }
 
   private void startServer() throws IOException {
-    startServer(100);
+    startServer(100, Http2CodecUtil.DEFAULT_MAX_HEADER_SIZE);
```
Comment (nmittler) [inline netty/src/test/java/io/grpc/netty/NettyClientTransportTest.java]: and here.

File: netty/src/main/java/io/grpc/netty/NettyClientTransport.java (hunk 3 above)
Comment (madongfly) [inline netty/src/main/java/io/grpc/netty/NettyClientTransport.java]: Here the header table size is not configurable in gRPC, so we shouldn't have a redundant constant for it.

File: netty/src/main/java/io/grpc/netty/NettyServerTransport.java (hunk 4 above)
Comment (madongfly) [inline netty/src/main/java/io/grpc/netty/NettyServerTransport.java]: Same.

**Comment by madongfly** [comment]
@nmittler Thanks for the catch, PTAL

File: netty/src/main/java/io/grpc/netty/NettyClientTransport.java (hunk 3 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/netty/NettyClientTransport.java]: Ah right  :)

File: netty/src/main/java/io/grpc/netty/NettyServerTransport.java (hunk 4 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/netty/NettyServerTransport.java]: +1

**Comment by nmittler** [comment]
@madongfly LGTM, thanks!

**Pull request merged**